xbmc/windowing/wayland/WinSystemWayland.cpp

```diff
@@ -464,60 +532,215 @@ bool CWinSystemWayland::SetFullScreen(bool fullScreen, RESOLUTION_INFO& res, boo
+{
+  return SetResolutionExternal(fullScreen, res);
+}
+
+void CWinSystemWayland::ApplySizeUpdate(SizeUpdateInformation update)
+{
+  if (update.bufferScaleChanged)
+  {
+    // Buffer scale must also match egl size configuration
+    ApplyBufferScale();
+  }
+  if (update.surfaceSizeChanged)
+  {
+    // Mark everything opaque so the compositor can render it faster
+    // Do it here so size always matches the configured egl surface
+    CLog::LogF(LOGDEBUG, "Setting opaque region size %dx%d", m_surfaceSize.Width(), m_surfaceSize.Height());
+    wayland::region_t opaqueRegion = m_compositor.create_region();
+    opaqueRegion.add(0, 0, m_surfaceSize.Width(), m_surfaceSize.Height());
+    m_surface.set_opaque_region(opaqueRegion);
+  }
+  if (update.configuredSizeChanged)
+  {
+    // Update window decoration state
+    m_windowDecorator->SetState(m_configuredSize, m_scale, m_shellSurfaceState);
+    m_shellSurface->SetWindowGeometry(m_windowDecorator->GetWindowGeometry());
+  }
+  // Set always, because of initialization order GL context has to keep track of
+  // whether the size changed. If we skip based on update.bufferSizeChanged here,
+  // GL context will never get its initial size set.
+  SetContextSize(m_bufferSize);
+  if (update.bufferSizeChanged)
+  {
+    if (m_shellSurfaceState.test(IShellSurface::STATE_RESIZING))
+    {
+      // Reloading the skin is costly and takes some time -> during resizing
+      // (when the size changes quite often), only reload after no movement for some time
+      if (m_resizeSkinReloadTimer.IsRunning())
+      {
+        m_resizeSkinReloadTimer.Restart();
+      }
+      else
+      {
+        m_resizeSkinReloadTimer.Start(500u, false);
+      }
+    }
+    else
+    {
+      // Always reload when not resizing
+      if (!m_inhibitSkinReload)
+      {
+        // FIXME This assumes that the resolution has already been set. Should
+        // be moved to some post-change callback when resolution setting is refactored.
+        g_application.ReloadSkin();
+      }
+    }
+  }
 }
 
+void CWinSystemWayland::ProcessMessages()
+{
+  Actor::Message* message{};
+  Actor::MessageHandle lastConfigureMessage;
+  int skippedConfigures = -1;
+  int newScale = m_scale;
+
+  while (m_protocol.ReceiveOutMessage(&message))
+  {
+    Actor::MessageHandle guard{message};
+    switch (message->signal)
+    {
+      case WinSystemWaylandProtocol::CONFIGURE:
+        // Do not directly process configures, get the last one queued:
+        // While resizing, the compositor will usually send a configure event
+        // each time the mouse moves without any throttling (i.e. multiple times
+        // per rendered frame).
+        // Going through all those and applying them would waste a lot of time when
+        // we already know that the size is not final and will change again anyway.
+        skippedConfigures++;
+        lastConfigureMessage = std::move(guard);
```

Comment:
the message is sent via deep copy. why do you preserve the message itself instead of copying its content?